source/main.coffee

```diff
@@ -201,42 +105,44 @@ export compile = (src, options) ->
 
 # logs = []
 makeCache = ->
-  caches = new Map
+  stateCache = new StateCache
+  getStateKey = null
 
   # stack = []
 
   events =
     enter: (ruleName, state) ->
-      cache = caches.get(ruleName)
-      if cache
-        if cache.has(state.pos)
-          # logs.push "".padStart(stack.length * 2, " ") + ruleName + ":" + state.pos + "💰"
-          result = cache.get(state.pos)
-          return {
-            cache: if result then { ...result }
-          }
+      return if uncacheable.has(ruleName)
+
+      key = [ruleName, state.pos, ...getStateKey()]
+
+      if stateCache.has(key)
+        # logs.push "".padStart(stack.length * 2, " ") + ruleName + ":" + state.pos + "💰"
+        result = stateCache.get(key)
```

Comment:
Looking at the implementation of `has` here, we're doubling our work. Maybe just call `get` and check for undefined?

Actually, I worry that the implementation of `has` isn't that good: it's not caching `undefined` results. Isn't that what we use to say that a rule didn't match? I'm worried we're currently not caching non-matches.

Reply:
Hmm, the `undefined` issue goes all the way to Hera. It generates:
```js
if (result.cache)
  return result.cache`
```
One solution would be to instead generate
```js
if ('cache' in result)
  return result.cache
```